web3.js/src/connection.ts

```diff
@@ -3085,9 +3116,66 @@ export class Connection {
           () => resolve({__type: TransactionStatus.TIMED_OUT, timeoutMs}),
           timeoutMs,
         );
+      } else if (Object.prototype.hasOwnProperty.call(strategy, 'nonceValue')) {
+        const INVALID_NONCE_ERROR_SIGIL = {} as const;
+        const {minContextSlot, nonceAccountPubkey, nonceValue} =
+          strategy as NonceBasedTransactionConfirmationStrategy;
+        let currentNonceValue = nonceValue;
+        const getCurrentNonceValue = async () => {
+          try {
+            const nonceAccount = await this.getNonce(nonceAccountPubkey, {
+              commitment: subscriptionCommitment,
+              minContextSlot,
+            });
+            if (nonceAccount == null) {
+              throw INVALID_NONCE_ERROR_SIGIL;
+            }
+            return nonceAccount.nonce;
+          } catch (e) {
+            if (e === INVALID_NONCE_ERROR_SIGIL) {
+              throw e;
+            }
+            // If for whatever reason we can't reach/read the nonce
+            // account, just keep using the last-known value.
+            return currentNonceValue;
+          }
+        };
+        (async () => {
+          try {
+            currentNonceValue = await getCurrentNonceValue();
+            if (done) return;
+            while (
+              true // eslint-disable-line no-constant-condition
+            ) {
```

Comment:
Could this be refactored to `while (!done)` and then each loop we either sleep or get the current nonce value? That way we don't need so many `done` checks.

Reply:
There’s a `done` check after _each_ async operation, because it’s at that point we need to check to make sure a confirmation didn’t come in while we were waiting for the async op to finish.

For instance, we want to check for done after the sleep, so that we don’t waste time on another nonce fetch for no reason. Putting the done check at the loop level would make this impossible.

Reply:
> Putting the done check at the loop level would make this impossible

My suggestion was to make the loop only run one async operation per loop so that done is still called immediately after. So it's possible, but I suppose it's less obvious so this isn't a big deal.